src/main/java/xyz/nucleoid/plasmid/game/stats/StatisticKey.java

```diff
@@ -0,0 +1,84 @@
+package xyz.nucleoid.plasmid.game.stats;
+
+import com.google.gson.JsonObject;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.StringIdentifiable;
+import xyz.nucleoid.plasmid.game.stats.key.DoubleStatisticKey;
+import xyz.nucleoid.plasmid.game.stats.key.FloatStatisticKey;
+import xyz.nucleoid.plasmid.game.stats.key.IntStatisticKey;
+
+/**
+ * Type-safe keys for identifying statistics.
+ * <p>
+ * Ideally should have a corresponding translation key in the form <code>statistic.[bundle namespace].[key namespace].[key path]</code>
+ * or <code>statistic.[key namespace].[key path]</code> for future-proofing
+ *
+ * @param <T> The type of {@link Number} this key stores
+ */
+public abstract class StatisticKey<T extends Number> {
+    protected final Identifier id;
+    protected final StatisticType type;
+
+    protected StatisticKey(Identifier id, StatisticType type) {
+        StatisticKey.validateKeyId(id);
+        this.id = id;
+        this.type = type;
+    }
+
+    public Identifier getId() {
+        return this.id;
+    }
+
+    @SuppressWarnings("unchecked")
+    public JsonObject encodeValueUnchecked(Number value) {
+        return this.encodeValue((T) value);
+    }
+
+    public JsonObject encodeValue(T value) {
+        JsonObject obj = new JsonObject();
+        obj.addProperty("type", this.encodeType());
+        obj.addProperty("value", value);
+        return obj;
+    }
+
+    protected abstract String encodeType();
```

Comment:
I think given this is the only abstract function here, this class probably doesn't need to be abstract- maybe it should instead hold two types like a data type (float/int/etc) and a store type (rolling avg/total)?